components/places/src/storage/bookmarks.rs

```diff
@@ -376,6 +395,259 @@ fn delete_bookmark_in_tx(db: &impl ConnExt, guid: &SyncGuid) -> Result<bool> {
+pub enum UpdateTreeLocation {
+    None,                               // no change
+    Position(BookmarkPosition),         // new position in the same folder.
+    Parent(SyncGuid, BookmarkPosition), // new parent
+}
+
+impl Default for UpdateTreeLocation {
+    fn default() -> Self {
+        UpdateTreeLocation::None
+    }
+}
+
+/// Structures which can be used to update a bookmark, folder or separator.
+/// Almost all fields are Option<>-like, with None meaning "do not change".
+/// Many fields which can't be changed by our public API are omitted (eg,
+/// guid, date_added, last_modified, etc)
+#[derive(Debug, Clone, Default)]
+pub struct UpdatableBookmark {
+    pub location: UpdateTreeLocation,
+    pub url: Option<Url>,
+    pub title: UpdateTitle,
+}
+
+impl From<UpdatableBookmark> for UpdatableItem {
+    fn from(bmk: UpdatableBookmark) -> Self {
+        UpdatableItem::Bookmark(bmk)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct UpdatableSeparator {
+    pub location: UpdateTreeLocation,
+}
+
+impl From<UpdatableSeparator> for UpdatableItem {
+    fn from(sep: UpdatableSeparator) -> Self {
+        UpdatableItem::Separator(sep)
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct UpdatableFolder {
+    pub location: UpdateTreeLocation,
+    // There's no good reason to differentiate `null` from `""` in a folder,
+    // but for consistency we allow it.
+    pub title: UpdateTitle,
+}
+
+impl From<UpdatableFolder> for UpdatableItem {
+    fn from(folder: UpdatableFolder) -> Self {
+        UpdatableItem::Folder(folder)
+    }
+}
+
+// The type used to update the actual item.
+#[derive(Debug, Clone)]
+pub enum UpdatableItem {
+    Bookmark(UpdatableBookmark),
+    Separator(UpdatableSeparator),
+    Folder(UpdatableFolder),
+}
+
+impl UpdatableItem {
+    fn bookmark_type(&self) -> BookmarkType {
+        match self {
+            UpdatableItem::Bookmark(_) => BookmarkType::Bookmark,
+            UpdatableItem::Separator(_) => BookmarkType::Separator,
+            UpdatableItem::Folder(_) => BookmarkType::Folder,
+        }
+    }
+
+    pub fn location(&self) -> &UpdateTreeLocation {
+        match self {
+            UpdatableItem::Bookmark(b) => &b.location,
+            UpdatableItem::Separator(s) => &s.location,
+            UpdatableItem::Folder(f) => &f.location,
+        }
+    }
+}
+
+pub fn update_bookmark(db: &impl ConnExt, guid: &SyncGuid, item: &UpdatableItem) -> Result<()> {
+    let tx = db.unchecked_transaction()?;
+    let result = update_bookmark_in_tx(db, guid, item);
+    match result {
+        Ok(_) => tx.commit()?,
+        Err(_) => tx.rollback()?,
+    }
+    result
+}
+
+fn update_bookmark_in_tx(db: &impl ConnExt, guid: &SyncGuid, item: &UpdatableItem) -> Result<()> {
+    let existing =
+        get_raw_bookmark(db, guid)?.ok_or_else(|| InvalidPlaceInfo::NoItem(guid.to_string()))?;
+    if existing.bookmark_type != item.bookmark_type() {
+        return Err(InvalidPlaceInfo::InvalidBookmarkType.into());
```

Comment:
Let's maybe give this error a more descriptive name, like `MismatchedType`, and specify the existing and new types in the message?